File: app/src/main/java/com/example/rajiv/dictionary/WordData.java
package com.example.rajiv.dictionary;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Observer;
import android.support.annotation.Nullable;

import com.example.rajiv.dictionary.definition.ResponseData;
import com.example.rajiv.dictionary.definition.Entry;
import com.example.rajiv.dictionary.definition.Sense;
import com.example.rajiv.dictionary.lemma.InflectionOf;
import com.example.rajiv.dictionary.lemma.Lemmatron;
import com.example.rajiv.dictionary.lemma.LexicalEntry;
import com.example.rajiv.dictionary.lemma.Result;
import com.example.rajiv.dictionary.pronunciation.Data;
import com.example.rajiv.dictionary.pronunciation.Pronunciation;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Retrofit;

import static com.example.rajiv.dictionary.RetrofitInstance.getRetrofitInstance;

public class WordData {
    private MutableLiveData<List<List<String>>> meanings;
    private  OxfordApiService service;


    public WordData(){
        meanings = new MutableLiveData<>();
        List<List<String>> lists = new ArrayList<>();
        meanings.setValue(lists);


    }

    private MutableLiveData<String> getLemma(String word){
        Retrofit retrofit = RetrofitInstance.getRetrofitInstance();
        OxfordApiService service = retrofit.create(OxfordApiService.class);

        final MutableLiveData<String> headWord = new MutableLiveData<>();

        Call<Lemmatron> result = service.getLemma(word);
        result.enqueue(new Callback<Lemmatron>(){
            @Override
            public void onResponse(Call<Lemmatron> call,
                                   retrofit2.Response<Lemmatron> response) {
                if (response.isSuccessful()) {
                    Lemmatron lemmatron = response.body();
                    List<Result> result = lemmatron.getResults();
                    List<LexicalEntry> lexicalEntry = result.get(0).getLexicalEntries();
                    List<InflectionOf> inflectionOf = lexicalEntry.get(0).getInflectionOf();
                    String lemma = inflectionOf.get(0).getText();
                    headWord.setValue(lemma);
                }
                else{
                    headWord.setValue("");
                }
            }

            public void onFailure(Call<Lemmatron> call, Throwable t){

            }
        });
        return headWord;
    }

    public MutableLiveData<List<List<String>>> getMeaning(String word){
        MutableLiveData<String> headWord = getLemma(word);
        headWord.observeForever(new Observer<String>() {
            @Override
            public void onChanged(@Nullable String lemma) {
                if (lemma != null) {
                    if (lemma.equals("")) {
                        meanings.setValue(null);

                    } else {

                        Retrofit retrofit = (new RetrofitInstance()).getRetrofitInstance();
                        OxfordApiService service = retrofit.create(OxfordApiService.class);

                        Call<ResponseData> result = service.getMeaning(lemma);
                        result.enqueue(new Callback<ResponseData>() {
                            @Override
                            public void onResponse(Call<ResponseData> call, retrofit2.Response<ResponseData> response) {
                                List<com.example.rajiv.dictionary.definition.Result> results;
                                List<com.example.rajiv.dictionary.definition.LexicalEntry> lexicalEntries =
                                        new ArrayList<>();
                                List<Entry> entries = new ArrayList<>();
                                List<Sense> senses = new ArrayList<>();
                                List<String> definitions = new ArrayList<>();
                                List<String> nounDefinitions = new ArrayList<>();
                                List<String> verbDefinitions = new ArrayList<>();
                                List<String> adjectiveDefinitions = new ArrayList<>();


                                ResponseData responseData = response.body();
                                results = responseData.getResults();

                                // Create LexicalEntry List
                                int resultLength = results.size();
                                for (int i = 0; i < resultLength; i++) {
                                    int lexicalSize = results.get(i).getLexicalEntries().size();
                                    for (int j = 0; j < lexicalSize; j++) {
                                        lexicalEntries.add(results.get(i).getLexicalEntries().get(j));
                                    }
                                }


                                for (com.example.rajiv.dictionary.definition.LexicalEntry l : lexicalEntries) {
                                    entries = l.getEntries();
                                    for (Entry e : entries) {
                                        senses = e.getSenses();
                                        for (Sense s : senses) {
                                            definitions = s.getDefinitions();
                                            String lexicalCategory = l.getLexicalCategory();
                                            for (String d : definitions) {
                                                if (lexicalCategory.equals("Noun") && d!=null) {
                                                    //for (String n : definitions)
                                                        nounDefinitions.add(d);
                                                }
                                                if (lexicalCategory.equals("Verb")) {
                                                    //for (String v : definitions)
                                                        verbDefinitions.add(d);
                                                }
                                                if (lexicalCategory.equals("Adjective")) {
                                                    //for (String a : definitions)
                                                        adjectiveDefinitions.add(d);
                                                }
                                            }
                                        }
                                    }
                                }

                                List<List<String>> meaning = new ArrayList<>();
                                meaning.add(nounDefinitions);
                                meaning.add(adjectiveDefinitions);
                                meaning.add(verbDefinitions);

                                meanings.setValue(meaning);
                            }

                            @Override
                            public void onFailure(Call<ResponseData> call, Throwable t) {


                            }
                        });
                    }

                }
            }
        });

        return meanings;

    }

    public MutableLiveData<String> getPronunciation(String word){
       Retrofit retrofit = RetrofitInstance.getRetrofitInstance();
       OxfordApiService service = retrofit.create(OxfordApiService.class);

       final MutableLiveData<String> pronuciation = new MutableLiveData<>();

       Call<Data> result = service.getPronunciation(word);
       result.enqueue(new Callback<Data>(){
           @Override
           public void onResponse(Call<Data> call,
                                  retrofit2.Response<Data> response) {
               if (response.isSuccessful()) {
                   Data data = response.body();
                   List<com.example.rajiv.dictionary.pronunciation.Result> result = data.getResults();
                   List<com.example.rajiv.dictionary.pronunciation.LexicalEntry> lexicalEntry = result.get(0).getLexicalEntries();
                   List<Pronunciation> pronunciations = lexicalEntry.get(0).getPronunciations();
                   String audioURL = pronunciations.get(0).getAudioFile();
                   pronuciation.setValue(audioURL);
               }
               else{
                   pronuciation.setValue("");
               }
           }

           public void onFailure(Call<Data> call, Throwable t){

           }
       });
       return pronuciation;


   }
}

File: app/src/main/java/com/example/rajiv/dictionary/WordData.kt
package com.example.rajiv.dictionary

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModel


import com.example.rajiv.dictionary.pronunciation.Data
import com.example.rajiv.dictionary.pronunciation.Pronunciation

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Retrofit

import com.example.rajiv.dictionary.RetrofitInstance.getRetrofitInstance
import com.example.rajiv.dictionary.definition.*
import com.example.rajiv.dictionary.lemma.GrammaticalFeature
import com.example.rajiv.dictionary.lemma.InflectionOf
import com.example.rajiv.dictionary.lemma.Lemmatron
import retrofit2.Response
import java.io.File
import java.util.*
import kotlin.collections.ArrayList

class WordData: ViewModel() {
    private var word: String? = null
    private var wordFullData: List<Result>? = null

    private lateinit var shortDef: MutableLiveData<ArrayList<String>>
    private lateinit var detailDef: MutableLiveData<ArrayList<DefWithCategory>>
    private lateinit var pronounceFile: File

    private val meanings: MutableLiveData<ArrayList<DefWithCategory>> = MutableLiveData()
    val retrofit = RetrofitInstance.getRetrofitInstance()
    val service: OxfordApiService? = retrofit.create(OxfordApiService::class.java)

    fun getWordFullData(inputWord: String) {
        val query = service!!.getFullWordData(inputWord)
        query.enqueue(object: Callback<ResponseData>{
            override fun onResponse(call: Call<ResponseData>?, response: retrofit2.Response<ResponseData>) {
                if(response.isSuccessful){
                    val responseData = response.body()
                    wordFullData = responseData!!.results
                }
                else{

                }
            }

            override fun onFailure(call: Call<ResponseData>?, t: Throwable?) {
                //TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }
        })
    }

    fun getShortDef(inputWord: String): MutableLiveData<ArrayList<String>> {
        if (inputWord.equals(word))
            return shortDef
        else{
            getWordFullData(inputWord)
        }
    }

    fun getMeaning(word: String): MutableLiveData<ArrayList<DefWithCategory>> {
                    val result = service!!.getFullWordData(word)
                    result?.enqueue(object : Callback<ResponseData> {
                        override fun onResponse(call: Call<ResponseData>, response: retrofit2.Response<ResponseData>) {
                            val results: List<Result>
                            val lexicalEntries = ArrayList<LexicalEntry>()
                            var entries: List<Entry> = ArrayList()
                            var senses: List<Sense> = ArrayList()
                            var definitions = ArrayList<DefWithCategory>()
                            val nounDefinitions = ArrayList<String>()
                            val verbDefinitions = ArrayList<String>()
                            val adjectiveDefinitions = ArrayList<String>()

                            if(response.isSuccessful) {
                                val responseData = response.body()

                                results = responseData!!.results
                                wordFullData = results

                                // Create LexicalEntry List
                                for (res in results) {
                                    res?.let {
                                        for (lex in it.lexicalEntries) {
                                            lex?.let { lexicalEntries.add(it) }
                                        }
                                    }
                                }

                                for (l in lexicalEntries) {
                                    for (e in l.entries) {
                                        e?.let {
                                            for (s in it.senses) {
                                                val definition = (s.definitions)
                                                        ?: s.crossReferenceMarkers
                                                definition?.let {
                                                    for (def in definition) {
                                                        val defWithCategory = DefWithCategory(def.capitalize()+".\n\n", l.lexicalCategory.capitalize())
                                                        definitions.add(defWithCategory)
                                                    }
                                                }

                                            }
                                        }
                                    }
                                }
                                val condensedDef = createObjList(definitions)
                                meanings.value = condensedDef
                            }
                            else{
                                val result = service!!.getLemma(word)
                                result?.enqueue(object: Callback<Lemmatron> {
                                    override fun onResponse(call: Call<Lemmatron>, response: Response<Lemmatron>){
                                        val results: List<com.example.rajiv.dictionary.lemma.Result>
                                        val lexical: com.example.rajiv.dictionary.lemma.LexicalEntry
                                        val grammar: GrammaticalFeature
                                        val inflection: InflectionOf

                                        if(response.isSuccessful){
                                            val responseData = response.body()
                                            results = responseData!!.results
                                            lexical = results!!.get(0).lexicalEntries[0]
                                            grammar = lexical.grammaticalFeatures[0]
                                            inflection = lexical.inflectionOf[0]

                                            val definition: String = grammar.text + " of " + inflection.text + ".\n\n"
                                            val cat: String = lexical.lexicalCategory
                                            val def = DefWithCategory(definition, cat)
                                            //val definitions = ArrayList<DefWithCategory>()
                                            definitions.add(def)
                                            val condensedDef = createObjList(definitions)
                                            meanings.value = condensedDef


                                        }
                                        else{
                                            definitions.add(DefWithCategory("Please check the spelling of the word entered.", "Error"))
                                            meanings.value = definitions
                                        }
                                    }

                                    override fun onFailure(call: Call<Lemmatron>?, t: Throwable?) {

                                    }
                                })
                            }


                        }

                        override fun onFailure(call: Call<ResponseData>, t: Throwable) {

                        }
                    })

        return meanings

    }

    fun getPronunciation(word: String): MutableLiveData<String> {
        val retrofit = RetrofitInstance.getRetrofitInstance()
        val service = retrofit.create(OxfordApiService::class.java)

        val pronuciation = MutableLiveData<String>()

        val result = service.getPronunciation(word)
        result.enqueue(object : Callback<Data> {
            override fun onResponse(call: Call<Data>,
                                    response: retrofit2.Response<Data>) {
                if (response.isSuccessful) {
                    val data = response.body()
                    val result = data!!.results
                    val lexicalEntry = result[0].lexicalEntries
                    val pronunciations = lexicalEntry[0].pronunciations
                    val audioURL = pronunciations[0].audioFile
                    pronuciation.setValue(audioURL)
                } else {
                    pronuciation.setValue("")
                }
            }

            override fun onFailure(call: Call<Data>, t: Throwable) {

            }
        })
        return pronuciation


    }

    fun createObjList(list: ArrayList<DefWithCategory>): ArrayList<DefWithCategory>{
        val defList = ArrayList<DefWithCategory>()
        defList.add(list[0])

        loop@ for(l in list.subList(1, list.size)){
            for(def in defList){
                if(def.category == l.category) {
                    def.def +=  l.def
                    continue@loop
                }
            }
            defList.add(DefWithCategory(l.def, l.category))
        }
        return defList
    }
}
